### PR TITLE
Padded SGM

### LINKED ISSUE #316: Implement Padded SGM algorithm
## functionality
Input Graphs G1 and G2 with respective vertex sets {1,...,n_1} and {1,...,n_2} and with respective adjacency matrices A and B, and suppose n_1<n_2, use the SGM function to the padded A and B, then return the restriction of permutation matrix to the vertices of G1.

The output can be compared with Figure 3 in paper seeded graph matching: https://arxiv.org/pdf/1209.0367.pdf.

### PR DESCRIPTION
**Pull request by asaadeldin11**
#### Reference Issues/PRs
Fixes #316 

#### What does this implement/fix? Explain your changes.
This PR brings padded SGM to graspologic, allowing users to pass in input matrices of different sizes to `GraphMatch`. This is an implementation of the padding scheme describing in [Seeded Graph Matching](https://arxiv.org/pdf/1209.0367.pdf) (section 2.5). Both schemes described (adopted and naive), are implemented here, allowing users the option to choose which they'd prefer to use.

#### Any other comments?
Proof of effectiveness can be found in the following [notebook](https://nbviewer.jupyter.org/github/asaadeldin11/nbview/blob/master/padded_sgm.ipynb), where Figure 3 from SGM is replicated.

### CODE REVIEW

**Comment by asaadeldin11** [comment]
Still TODO: add tests

**Comment by asaadeldin11** [comment]
Also: do we want to use this PR to bring new scipy version of QAP into graspologic? My thought was to add it in as a new file until scipy's new release, get the formatting ready for when that happens, then when it does happen just change the import statement.

**Comment by bdpedigo** [comment]
> Also: do we want to use this PR to bring new scipy version of QAP into graspologic?

My perspective is each PR should be as atomic/small as possible, makes review way way easier. So I'd say keep separate as if it makes sense to do so

File: graspy/match/gmp.py [hunk 1]
```diff
@@ -57,6 +57,19 @@ class GraphMatch(BaseEstimator):
...
+
+        "adopted" : say that `A` and `B` have :math:`n_1` and :math:`n_2` nodes, respectively.
+        Define :math:`\widetilde{A} = 2A - 1_{n1}1_{n1}^T` and
+        :math:`\widetilde{B} = 2B - 1_{n2}1_{n2}^T`. If :math:`n_1 < n_2`, then we substitute
+        :math:`\widetilde{A} \oplus 0_{(n2-n1)x(n2-n1)}` and :math:`\widetilde{B}` in place of A
+        and B.
+
+        "naive" : say that `A` and `B` have :math:`n_1` and :math:`n_2` nodes, respectively.
+        If :math:`n_1 < n_2`, then we substitute :math:`A \oplus 0_{(n2-n1)x(n2-n1)}` and
+        :math:`B` in place of A and B.
```
Comment (bdpedigo) [inline graspy/match/gmp.py:71]: i don't know if this is the appropriate place for the math or not. regardless, none of this helps the user understand what to use and when

**Review by bdpedigo — CHANGES_REQUESTED** [review]
- looks like you also took out some testing in this PR? 
- see comments about docs, i don't think they are informative to a casual user ATM
- are you going to add something to tutorials? could either be a new notebook or maybe it makes more sense to just tack it on to what we have. if you don't do it here, im okay with it, but please make a new issue specifying that we want a tutorial, and also document what that tutorial should have (since you basically already have all the code)

**Comment by asaadeldin11** [comment]
> * are you going to add something to tutorials? could either be a new notebook or maybe it makes more sense to just tack it on to what we have. if you don't do it here, im okay with it, but please make a new issue specifying that we want a tutorial, and also document what that tutorial should have (since you basically already have all the code)

I think a good tutorial would be the recreation of figure 3 from the jupyter notebook. Shows the difference between naive and adopted padding as well.

File: graspy/match/gmp.py (hunk 1 above)
Comment (asaadeldin11) [reply graspy/match/gmp.py:71]: Makes sense. I'll add this math part to the tutorial.

**Comment by bdpedigo** [comment]
> looks like you also took out some testing in this PR?

please update PR desc accordingly

**Comment by bdpedigo** [comment]
@asaadeldin11 looks like this is from a fork, can you make this from a branch in main repo like we talked about yesterday (for build checks to work properly)

File: graspy/match/gmp.py [hunk 2]
```diff
@@ -126,6 +138,11 @@ def __init__(
         else:
             msg = '"gmp" must be a boolean'
             raise TypeError(msg)
+        if padding in {"adopted", "naive"}:
+            self.padding = padding
+        else:
+            msg = 'Invalid "padding" parameter string'
+            raise ValueError(msg)
```
Comment (bdpedigo) [inline graspy/match/gmp.py:145]: do you check that it is a string?

Comment (bdpedigo) [reply graspy/match/gmp.py:145]: and add corresponding test

File: tests/match/test_match.py [hunk 3]
```diff
@@ -137,18 +102,38 @@ def test_rand_SGM(self):
...
+                    stepx : (stepx + 75), stepy : (stepy + 75)
+                ]
+        seed1 = np.random.randint(0, 300, 10)
+        seed2 = [int(x / 75) * 25 + x for x in seed1]
+        res = self.barygm.fit(G2, G1, seed2, seed1)
+        matching = np.concatenate(
+            [res.perm_inds_[x * 100 : (x * 100) + 75] for x in range(n_blocks)]
+        )
+
+        assert 1.0 == (sum(matching == np.arange(300)) / 300)
```
Comment (bdpedigo) [inline tests/match/test_match.py:137]: i have a pretty hard time reading this test at a glance. 

could you add some comments explaining what is going on here, or better yet, just make the code simpler? 

I'm wondering if we also need this extensive/specific of a test? seems like super small rho-ER  with rho very high or 1, delete 1 or 2 vertices, show that the matching still works for the vertices that were not deleted in either graph. or something along those lines. not sure if we'd need one for adopted and one for naive, or maybe on a super simple example the same padding scheme would get the same answer? but we could test both

**Comment by bdpedigo** [comment]
Notebook comments: 
- remove the pylab thingy, not necessary right?
- "isolated nodes" is somewhat true, but they are also "empty row/columns in the adjacency matrices". Do you think that adds anything? somehow it makes more sense to me. just a question/comment tho
- "Since the isolated vertices and non adjacent core vertices both have the same weight, the isolated vertices carry a lot of false signal." I don't follow
- new paragraph after the sentence above.
- "Instead, we desire to match 𝐴 to the best fitting induced subgraph of 𝐵. " -> we may desire to match 𝐴...
- the two lines of import in the middle of the text explanation is weird to me
```c = np.copy(G1p)
keep = [np.random.randint(100*i, 100*(i+1)-1, 75) for i in range(4)]
delete = [np.setdiff1d(np.arange(100*i, 100*(i+1)-1), keep[i]) for i in range(4)]

for i in range(5):
    stepx = 100*i
    c[75+stepx:stepx+100,:] = -1
    c[:, 75+stepx:stepx+100] = -1
    for j in range(5):
        stepy = 100*j
        G1[(75*i):(75*(i+1)),(75*j):(75*(j+1))] = G1p[stepx:(stepx+75),stepy:(stepy+75)]

topleft_G1 = np.zeros((400,400))
topleft_G1[:300,:300] = G1```
- i think we can clean the above code up a lot, lets talk bout it
- Do we want to compute matching accuracy and show that? It seems like you were saying "adopted is better" and i was like... well is it?

**Review by bdpedigo — CHANGES_REQUESTED** [review]
(forgot to request changes, see above)

**Comment by asaadeldin11** [comment]
this PR superseded by #510

**Pull request closed without merging**